src/libraries/Common/src/Internal/Cryptography/UniversalCryptoOneShot.cs

```diff
@@ -19,27 +19,49 @@ internal static class UniversalCryptoOneShot
             if (input.Length % cipher.PaddingSizeInBytes != 0)
                 throw new CryptographicException(SR.Cryptography_PartialBlock);
 
-            // If there is no padding that needs to be removed, and the output buffer is large enough to hold
-            // the resulting plaintext, we can decrypt directly in to the output buffer.
-            // We do not do this for modes that require padding removal.
-            //
-            // This is not done for padded ciphertexts because we don't know if the padding is valid
-            // until it's been decrypted. We don't want to decrypt in to a user-supplied buffer and then throw
-            // a padding exception after we've already filled the user buffer with plaintext. We should only
-            // release the plaintext to the caller once we know the padding is valid.
-            if (!SymmetricPadding.DepaddingRequired(paddingMode))
+            // The internal implementation of the one-shots are never expected to create
+            // a plaintext larger than the ciphertext. If the buffer supplied is large enough
+            // to do the transform, use it directly.
+            // This does mean that the TransformFinal will write more than what is reported in
+            // bytesWritten when padding needs to be removed. The padding is "removed" simply
+            // by reporting less of the amount written, then zeroing out the extra padding.
+            if (output.Length >= input.Length)
             {
-                if (output.Length >= input.Length)
+                int bytesTransformed = cipher.TransformFinal(input, output);
+                Span<byte> transformBuffer = output.Slice(0, bytesTransformed);
+
+                try
                 {
-                    bytesWritten = cipher.TransformFinal(input, output);
+                    // validates padding
+                    // This intentionally passes in BlockSizeInBytes instead of PaddingSizeInBytes. This is so that
+                    // "extra padded" CFB data can still be decrypted. The .NET Framework always padded CFB8 to the
+                    // block size, not the feedback size. We want the one-shot to be able to continue to decrypt
+                    // those ciphertexts, so for CFB8 we are more lenient on the number of allowed padding bytes.
+                    bytesWritten = SymmetricPadding.GetPaddingLength(transformBuffer, paddingMode, cipher.BlockSizeInBytes);
+
+                    // Zero out the padding so that the buffer does not contain the padding data "after" the bytesWritten.
+                    CryptographicOperations.ZeroMemory(transformBuffer.Slice(bytesWritten));
                     return true;
                 }
+                catch (CryptographicException)
+                {
+                    // The padding is invalid, but don't leave the plaintext in the buffer.
+                    CryptographicOperations.ZeroMemory(transformBuffer);
```

Comment:
Once we're willing to let "unverified" data be written into the destination buffer, we could do all of this differently.

* If output.Length >= input.Length, do this.
* else if no removable padding, fail
* else if output.Length < input.Length - BlockSizeInBytes, fail
* try
  * Decrypt input.Length - BlockSizeInBytes into output.
  * stackalloc 16 (Debug.Assert BlockSizeInBytes <= stackalloc.Length)
  * Decrypt the last block into the stackalloc
  * Apply depadding, copy or fail as appropriate
* catch
  * clear output
  * throw-continue

And that eliminates the rented array entirely, at the cost of a second P/Invoke for the "it might or might not fit" case.

Historically, I'm not a fan of using caller provided buffers as scratch space; but it happens in enough places that I may as well embrace it.

Reply:
Hm. This gets a little ugly with CFB but, I think I can make it work.